# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [110]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [111]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [112]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [113]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [114]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [115]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [116]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [117]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [118]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [119]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [120]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [121]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [122]:
df['Season'].value_counts()

Season
2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: count, dtype: int64

**Q1**: How many unique seasons are there in the dataset? Use the nunique() function.

In [123]:
# Write your code here
# I use the given function "nunique()", and find the number of unique seasons, which is 32
df['Season'].nunique()

32

**Q2**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [124]:
# Write your code here
# I use the given function "value_counts()", and find the team with the most wins, which is 1181
df['Wteam'].value_counts().idxmax()

1181

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [125]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [126]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [127]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [128]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [129]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [130]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [131]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [132]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [133]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [134]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q3**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

To create the three dataframes sorted by season, winning team, and winning score, I start by using iloc. When using iloc, an extra value must be included; therefore, I use 100:201. This is applied to all three dataframes.

In [135]:
# Write your code here

# Sorted by season
sorted_season = df.sort_values('Season')
dataframe_season = sorted_season.iloc[100:201]
dataframe_season

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
2525    1985     100   1406      68   1137      65    H      0
2526    1985     100   1410      83   1320      59    H      0
2527    1985     100   1423      58   1187      56    H      0
2528    1985     100   1429      86   1253      59    H      0
2529    1985     100   1440      50   1456      48    H      0
...      ...     ...    ...     ...    ...     ...  ...    ...
2447    1985      98   1212      63   1398      53    H      0
2448    1985      98   1216      64   1134      63    H      0
2449    1985      98   1233      77   1264      69    H      0
2450    1985      98   1249      80   1427      70    H      0
2425    1985      96   1449      72   1113      65    A      0

[101 rows x 8 columns]

In [136]:
# Sorted by winning team
sorted_wteam = df.sort_values('Wteam')
dataframe_wteam = sorted_wteam.iloc[100:201]
dataframe_wteam

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
97792     2008       5   1102      71   1294      62    H      0
97721     2008       4   1102      67   1171      38    H      0
96770     2007     110   1102      67   1161      58    H      0
96658     2007     107   1102      69   1428      43    H      0
96454     2007     103   1102      60   1307      51    A      0
...        ...     ...    ...     ...    ...     ...  ...    ...
33628     1993      82   1102      71   1201      51    A      0
32738     1993      61   1102      69   1161      58    H      0
135550    2015      30   1102      59   1212      34    H      0
32610     1993      57   1102      75   1351      66    N      0
32310     1993      47   1102      96   1339      65    H      0

[101 rows x 8 columns]

In [137]:
# Sorted by winning score
sorted_wscore = df.sort_values('Wscore')
dataframe_wscore = sorted_wscore.iloc[100:201]
dataframe_wscore

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
99013     2008      33   1460      43   1139      42    H      0
126840    2013      77   1366      43   1126      40    H      0
126982    2013      80   1442      43   1236      40    A      0
15619     1989      30   1343      43   1159      33    H      0
122700    2012     103   1349      43   1374      39    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
135433    2015      26   1438      45   1353      26    N      0
72338     2002      90   1443      45   1117      42    H      0
135808    2015      36   1185      45   1276      42    A      0
129196    2013     131   1380      45   1341      44    N      0
132791    2014      92   1155      45   1210      41    H      0

[101 rows x 8 columns]

**Q4**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

To find the season and winning team with the highest winning score, I combine the three subsets created in Q3 and use the "idxmax()" function. Then, I extract the season, team, and score.

In [138]:
# Combine the three subset from Q3, with the function "concat()", which is used to combine multiple DataFrames along a specified axis.
combined_subset = pd.concat([dataframe_season, dataframe_wteam, dataframe_wscore])

# Find the game with the highest score, with "idxmax()"
highest_score_row = combined_subset.loc[combined_subset['Wscore'].idxmax()]

season = highest_score_row['Season']
winning_team = highest_score_row['Wteam']
winning_score = highest_score_row['Wscore']

print(f"The season is {season} and the winning team is {winning_team} with the winning score of {winning_score}.")

The season is 1985 and the winning team is 1260 with the winning score of 105.


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [139]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [140]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q5**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

The goal of this task is to create a new column in the dataframe called 'ScoreDifference', which represents the absolute difference between the winning team's score and the losing team's score.

First, the new column is created by taking the absolute value of winning score minus losing score.
Then, the average of 'ScoreDifference' is calculated using the "mean()" function.
Finally, a filter is applied to keep only the games where 'ScoreDifference' is greater than the average 'ScoreDifference'.

In [141]:
# Write your code here

# Create the new column 'ScoreDifference'
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])

# Calculate the average of 'ScoreDifference'
average_score_difference = df['ScoreDifference'].mean()

# Make the filter in the Dataframe so it only includes games where 'ScoreDifference' is bigger than average
filter_df = df[df['ScoreDifference'] > average_score_difference]

filter_df


Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
3         1985      25   1165      70   1432      54    H      0   
6         1985      25   1228      64   1226      44    N      0   
8         1985      25   1260      98   1133      80    H      0   
10        1985      25   1307     103   1288      71    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145280    2016     131   1401      71   1261      38    N      0   
145282    2016     131   1433      76   1172      54    N      0   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference  
0                    17  
3                    16  
6                    20  
8                    18  
10                   32  
...                 ...  
145280               33  
145282               22  
145284               20  
145285               14  
145288               13  

[57227 rows x 9 columns]

**Q6**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

To find the season and the teams involved in the game with the largest 'scoredifference', I follow these steps:

Identify the row with the highest 'scoredifference' using the "idxmax()" function.
Extract the corresponding season, winning team, losing team, and scoredifference from that row.

In [142]:
# Write your code here
# Find the row with the highest 'scoredifference' using the idxmax() function.
highest_diff_row = filter_df.loc[filter_df['ScoreDifference'].idxmax()]

season = highest_diff_row['Season']
winning_team = highest_diff_row['Wteam']
losing_team = highest_diff_row['Lteam']

print(f"The season is {season}, with the winning team {winning_team}, and the losing team {losing_team}.")

The season is 1996, with the winning team 1409, and the losing team 1341.


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [143]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [144]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: count, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [145]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [146]:
df.values[0][0]

1985

**Q7**: Group the DataFrame by season and find the average winning score for each season.

To group the DataFrame by season and calculate the average winning score, I follow these steps:

I use the "groupby()" function to group the data by season. Then select the winning score column.
And lastly, apply the "mean()" function to compute the average winning score for each season.

In [147]:
# Write your code here
average_winning_scores_by_season = df.groupby('Season')['Wscore'].mean()

average_winning_scores_by_season

Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958
Name: Wscore, dtype: float64

**Q8**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

To find the maximum winning score for each team across all seasons, similar to Q7, follow these steps:

I use "groupby()" on 'winning team' and select the 'winning score' column. Lastly, I apply the "max()" function to get the highest winning score for each team.

In [148]:
# Write your code here
max_winning_scores_by_team = df.groupby('Wteam')['Wscore'].max()

max_winning_scores_by_team

Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
       ... 
1460    136
1461    112
1462    125
1463    105
1464    115
Name: Wscore, Length: 364, dtype: int64

**Q9**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

I start by grouping by season and winning team and calculating the average. Then, I create a new DataFrame and add the column 'Averagewinningscore'. Finally, I identify the team with the highest average winning score using "idxmax()".

To find the team with the highest average I use the "max()"

In [149]:
# Write your code here
# I start by grouping by season and winning team and calculating the average
avg_winning_scores = df.groupby(['Season', 'Wteam'])['Wscore'].mean()

# Make a new dataframe 
avg_winning_scores_df = avg_winning_scores.reset_index()
avg_winning_scores_df.columns = ['Season', 'Wteam', 'AverageWinningScore']

highest_avg_scores_per_season = avg_winning_scores_df.loc[avg_winning_scores_df.groupby('Season')['AverageWinningScore'].idxmax()]

highest_avg_scores_per_season

Season  Wteam  AverageWinningScore
174     1985   1328            92.800000
287     1986   1109            91.200000
783     1987   1380            95.875000
977     1988   1258           111.750000
1268    1989   1258           117.315789
1560    1990   1258           126.347826
1950    1991   1380           112.312500
2246    1992   1380            99.642857
2542    1993   1380           101.875000
2842    1994   1380           106.583333
3003    1995   1206           102.833333
3305    1996   1206           103.600000
3648    1997   1254            94.142857
4070    1998   1395           101.153846
4312    1999   1317            95.750000
4695    2000   1395            95.647059
5013    2001   1395            99.315789
5201    2002   1242            92.888889
5660    2003   1395            88.222222
5809    2004   1194            92.142857
6239    2005   1311            91.333333
6460    2006   1190            89.533333
7026    2007   1440            99.400000
7367    2008   1440            94.500000
7652    2009   1377            96.200000
8061    2010   1440            95.000000
8414    2011   1449            90.130435
8683    2012   1370            87.400000
9036    2013   1377            86.500000
9333    2014   1322            92.600000
9683    2015   1322            90.733333
9861    2016   1146            97.142857

In [150]:
# Find the overall highest average winning score across all seasons
overall_highest_avg_score = highest_avg_scores_per_season['AverageWinningScore'].max()

# Find the row corresponding to the highest score
highest_score = highest_avg_scores_per_season.loc[highest_avg_scores_per_season['AverageWinningScore'] == overall_highest_avg_score]
highest_score


Season  Wteam  AverageWinningScore
1560    1990   1258           126.347826

**Q10**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

I create a new DataFrame "Wins_per_team_season" to be able to count the number of wins for each season. 

In [151]:
wins_per_team_season = df.groupby(['Season', 'Wteam'])['Wscore'].count()
wins_per_team_season

Season  Wteam
1985    1102      5
        1103      9
        1104     21
        1106     10
        1108     19
                 ..
2016    1460     20
        1461     12
        1462     27
        1463     21
        1464      9
Name: Wscore, Length: 10172, dtype: int64

**Q11**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

To find which team has the most wins in each season, I count the number of wins for each team in every season. Then, I use "idxmax()" to find the team with the most wins.

In [152]:
# Write your code here
# I create a dataframe 
wins_by_team_season = df.groupby(['Season', 'Wteam']).size().reset_index(name='Wins')

# Find the team with the most wins in each season
most_wins_per_season = wins_by_team_season.loc[wins_by_team_season.groupby('Season')['Wins'].idxmax()]
most_wins_per_season


Season  Wteam  Wins
217     1985   1385    27
342     1986   1181    32
818     1987   1424    33
861     1988   1112    31
1323    1989   1328    28
1551    1990   1247    29
1739    1991   1116    30
2088    1992   1181    28
2423    1993   1231    28
2665    1994   1163    27
2928    1995   1116    26
3357    1996   1269    31
3638    1997   1242    31
3943    1998   1242    33
4201    1999   1181    32
4709    2000   1409    29
4829    2001   1181    29
5121    2002   1153    30
5456    2003   1166    29
5981    2004   1390    29
6165    2005   1228    32
6452    2006   1181    30
6841    2007   1242    30
7209    2008   1272    33
7552    2009   1272    31
7870    2010   1242    32
8217    2011   1242    32
8565    2012   1246    32
8878    2013   1211    30
9462    2014   1455    33
9609    2015   1246    34
9955    2016   1242    29

**Q12**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [153]:
# Write your code here
# Calculate the average losing score for each team
avg_losing_scores = df.groupby('Lteam')['Lscore'].mean().reset_index(name='AvgLosingScore')

# Calculate the average winning score for each team 
avg_winning_scores = df.groupby('Wteam')['Wscore'].mean().reset_index(name='AvgWinningScore')

# Merge the data to compare
comparison_df = pd.merge(avg_winning_scores, avg_losing_scores, left_on='Wteam', right_on='Lteam', how='outer')

# find teams with a higher average losingscore than winningscore
teams_higher_losing_than_winning = comparison_df[comparison_df['AvgLosingScore'] > comparison_df['AvgWinningScore']]

if teams_higher_losing_than_winning.empty:
    print("There are no teams with a higher average losing score than their average winning score.")
else:
    print("Teams with a higher average losing score than average winning score:")
    print(teams_higher_losing_than_winning)


There are no teams with a higher average losing score than their average winning score.


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [154]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q13**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

I add a new column 'HighScoringGame', which is set to "Yes" if the winning score is greater than 100 and "No" otherwise. Then, I check this condition for all rows in the DataFrame.

In [155]:
# Write your code here
df['HighScoringGame'] = df['Wscore'].apply(lambda x: 'Yes' if x > 100 else 'No')

df[['Wscore', 'HighScoringGame']]

Wscore HighScoringGame
0           81              No
1           77              No
2           63              No
3           70              No
4           86              No
...        ...             ...
145284      70              No
145285      72              No
145286      82              No
145287      66              No
145288      87              No

[145289 rows x 2 columns]

**Q14**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

I made a code that calculates the total number of games played by each team, whether they won or lost.

Initialization: An empty dictionary (team_game_counts) is created to keep track of the number of games for each team.

Iterating through the DataFrame: The "iterrows()" function is used to loop through all rows in the DataFrame.

Updating game counts:

If a team is already in the dictionary, its count is incremented by 1.

If not, the team is added with an initial count of 1.

Result: The dictionary team_game_counts displays the total number of games played by each team.

In [156]:
# Write your code here
# Initialize an empty dictionary to keep track of game counts
games_played = {}

# Iterate over the DataFrame rows
for index, row in df.iterrows():
    # Update counts for the winning team
    if row['Wteam'] in games_played:
        games_played[row['Wteam']] += 1
    else:
        games_played[row['Wteam']] = 1
    
    # Update counts for the losing team
    if row['Lteam'] in games_played:
        games_played[row['Lteam']] += 1
    else:
        games_played[row['Lteam']] = 1

games_played

{1228: 992,
 1328: 968,
 1106: 855,
 1354: 906,
 1112: 981,
 1223: 363,
 1165: 833,
 1432: 69,
 1192: 908,
 1447: 903,
 1218: 931,
 1337: 922,
 1226: 847,
 1242: 993,
 1268: 969,
 1260: 914,
 1133: 949,
 1305: 922,
 1424: 974,
 1307: 969,
 1288: 925,
 1344: 951,
 1438: 952,
 1374: 916,
 1411: 903,
 1412: 962,
 1397: 963,
 1417: 966,
 1225: 880,
 1116: 980,
 1368: 808,
 1120: 936,
 1391: 879,
 1135: 847,
 1306: 898,
 1143: 947,
 1388: 897,
 1153: 970,
 1184: 863,
 1159: 887,
 1171: 829,
 1216: 930,
 1173: 960,
 1134: 200,
 1177: 942,
 1296: 879,
 1193: 942,
 1265: 934,
 1196: 981,
 1416: 881,
 1206: 938,
 1137: 912,
 1210: 972,
 1149: 824,
 1211: 921,
 1102: 840,
 1234: 968,
 1114: 910,
 1332: 927,
 1243: 927,
 1317: 883,
 1257: 994,
 1231: 969,
 1277: 966,
 1145: 934,
 1278: 948,
 1453: 912,
 1286: 851,
 1186: 849,
 1301: 979,
 1144: 850,
 1325: 942,
 1384: 887,
 1326: 968,
 1248: 896,
 1287: 857,
 1339: 879,
 1334: 899,
 1365: 907,
 1375: 896,
 1126: 906,
 1403: 939,
 1152: 865,
 1423

**Q15**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

This code finds the game with the highest score difference (winning score - losing score) for each season.

Initialize a dictionary: season_highest_difference is created to store the highest score difference for each season.

Iterate through the DataFrame: The code loops through each row using "iterrows()".

Calculate the score difference: The difference between the winning and losing score is computed.

Update the dictionary:

If the season is not in the dictionary, the current game is stored as the highest difference.

If the season already exists, the game is updated only if the current game has a higher score difference.

Convert to DataFrame: The dictionary is transformed into a DataFrame

Sorting: The final DataFrame is sorted by season for clarity.

In [157]:
# Initialize a dictionary to track the highest score difference for each season
season_highest_difference = {}

# Iterate through each row of the DataFrame
for index, row in df.iterrows():
    # Calculate the score difference
    score_difference = row['Wscore'] - row['Lscore']
    
    # Get the current season
    season = row['Season']
    
    # Check if this season is already in the dictionary
    if season not in season_highest_difference:
        # If not, initialize with the current game
        season_highest_difference[season] = {
            'ScoreDifference': score_difference,
            'WinningTeam': row['Wteam'],
            'LosingTeam': row['Lteam'],
            'WinningScore': row['Wscore'],
            'LosingScore': row['Lscore']
        }
    else:
        # If it exists, check if the current game has a higher score difference
        if score_difference > season_highest_difference[season]['ScoreDifference']:
            season_highest_difference[season] = {
                'ScoreDifference': score_difference,
                'WinningTeam': row['Wteam'],
                'LosingTeam': row['Lteam'],
                'WinningScore': row['Wscore'],
                'LosingScore': row['Lscore']
            }

# Convert the dictionary to a DataFrame for better presentation
highest_diff_per_season = pd.DataFrame.from_dict(season_highest_difference, orient='index').reset_index()
highest_diff_per_season.columns = ['Season', 'ScoreDifference', 'WinningTeam', 'LosingTeam', 'WinningScore', 'LosingScore']

# Sort by season for better readability
highest_diff_per_season = highest_diff_per_season.sort_values(by='Season')

highest_diff_per_season


Season  ScoreDifference  WinningTeam  LosingTeam  WinningScore  \
0     1985               60         1361        1288           128   
1     1986               84         1314        1264           129   
2     1987               73         1155        1118           112   
3     1988               68         1328        1147           152   
4     1989               70         1242        1135           115   
5     1990               76         1181        1217           130   
6     1991               68         1163        1148           115   
7     1992               82         1116        1126           128   
8     1993               81         1328        1197           146   
9     1994               69         1228        1152           121   
10    1995               74         1246        1404           124   
11    1996               91         1409        1341           141   
12    1997               80         1278        1106           114   
13    1998               66         1395        1410           153   
14    1999               75         1268        1317           132   
15    2000               75         1261        1212           112   
16    2001               75         1196        1197           125   
17    2002               59         1328        1183           109   
18    2003               72         1403        1311           107   
19    2004               70         1328        1115            94   
20    2005               64         1403        1421           119   
21    2006               72         1400        1341           110   
22    2007               74         1401        1212           101   
23    2008               70         1182        1224           129   
24    2009               68         1314        1421           116   
25    2010               75         1397        1421           124   
26    2011               69         1388        1152           121   
27    2012               62         1345        1296            96   
28    2013               58         1235        1105            98   
29    2014               74         1256        1255           126   
30    2015               69         1181        1342           113   
31    2016               73         1139        1154           144   

    LosingScore  
0            68  
1            45  
2            39  
3            84  
4            45  
5            54  
6            47  
7            46  
8            65  
9            52  
10           50  
11           50  
12           34  
13           87  
14           57  
15           37  
16           50  
17           50  
18           35  
19           24  
20           55  
21           38  
22           27  
23           59  
24           48  
25           49  
26           52  
27           34  
28           40  
29           52  
30           44  
31           71

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [158]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q16**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

First, the number of wins and losses is counted using "value_counts()". Then, the results are combined using "add()", with missing values set to 0. The data is then converted into a DataFrame, and finally, it is sorted by the total number of games played.

In [159]:
# Write your code here

# Calculate the number of games won by each team
winning_team_counts = df['Wteam'].value_counts()

# Calculate the number of games lost by each team
losing_team_counts = df['Lteam'].value_counts()

# Combine the two Series to get the total games played by each team
total_games_played = winning_team_counts.add(losing_team_counts, fill_value=0).astype(int)

# Convert to DataFrame for better presentation
total_games_played_df = total_games_played.reset_index()
total_games_played_df.columns = ['Team', 'TotalGames']

# Sort by total games played
total_games_played_df = total_games_played_df.sort_values(by='TotalGames', ascending=False)

total_games_played_df


Team  TotalGames
80   1181        1013
213  1314        1010
145  1246         999
156  1257         994
292  1393         993
..    ...         ...
129  1230          64
345  1446          48
188  1289          48
17   1118          27
226  1327          22

[364 rows x 2 columns]

**Q17**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

I find the game with the highest score difference (winning score - losing score) for each season using the vectorized approach.

I create a new column called 'ScoreDifference', where the losing score is subtracted from the winning score.
Then, the data is grouped by season, and the "idxmax()" function is used to find the index of the game with the largest score difference in each season.
Finally, the row with the highest score difference per season is selected using "loc()".

In [160]:
# Write your code here
# Create the new column 'ScoreDifference' using vectorized subtraction
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

# Finding the index of the game with the highest score difference for each season
highest_diff_idx = df.groupby('Season')['ScoreDifference'].idxmax()

# Use the indices to extract the rows with the highest score difference for each season
highest_diff_games = df.loc[highest_diff_idx]

highest_diff_games[['Season', 'Wteam', 'Lteam', 'Wscore', 'Lscore', 'ScoreDifference']]


Season  Wteam  Lteam  Wscore  Lscore  ScoreDifference
236       1985   1361   1288     128      68               60
4731      1986   1314   1264     129      45               84
8240      1987   1155   1118     112      39               73
12046     1988   1328   1147     152      84               68
16677     1989   1242   1135     115      45               70
19502     1990   1181   1217     130      54               76
25161     1991   1163   1148     115      47               68
27997     1992   1116   1126     128      46               82
33858     1993   1328   1197     146      65               81
36404     1994   1228   1152     121      52               69
39858     1995   1246   1404     124      50               74
44653     1996   1409   1341     141      50               91
49033     1997   1278   1106     114      34               80
52600     1998   1395   1410     153      87               66
57618     1999   1268   1317     132      57               75
60616     2000   1261   1212     112      37               75
65999     2001   1196   1197     125      50               75
70737     2002   1328   1183     109      50               59
74791     2003   1403   1311     107      35               72
79181     2004   1328   1115      94      24               70
83372     2005   1403   1421     119      55               64
89542     2006   1400   1341     110      38               72
94507     2007   1401   1212     101      27               74
97737     2008   1182   1224     129      59               70
103723    2009   1314   1421     116      48               68
108405    2010   1397   1421     124      49               75
114015    2011   1388   1152     121      52               69
118699    2012   1345   1296      96      34               62
124057    2013   1235   1105      98      40               58
131373    2014   1256   1255     126      52               74
134590    2015   1181   1342     113      44               69
140040    2016   1139   1154     144      71               73

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [161]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [162]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [163]:
type(df['Wscore'])

pandas.core.series.Series

In [164]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [165]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [166]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [167]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDifference    0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y